api_app/service_bus/substitutions.py

```diff
@@ -89,18 +103,54 @@ def substitute_value(val: str, primary_resource_dict: dict) -> Union[dict, list,
             t = p[0:p.index("}}")]
             tokens.append(t)
 
+    dict_to_use = None
     for t in tokens:
-        # t = "resource.properties.prop_1"
+        # t = "{resource[.parent][.parent].properties.prop_1"
         p = t.split(".")
-        if p[0] == "resource":
-            prop_to_get = primary_resource_dict
-            for i in range(1, len(p)):
-                prop_to_get = prop_to_get[p[i]]
-
-            # if the value to inject is actually an object / list - just return it, else replace the value in the string
-            if isinstance(prop_to_get, dict) or isinstance(prop_to_get, list):
-                return prop_to_get
-            else:
-                val = val.replace("{{" + t + "}}", str(prop_to_get))
+
+        # decide on which dictionary to use (parents support)
+        # how many parents levels do we have (0- current resource, 1-direct parent, 2-skip level parent, 3-invalid)
+        hierarchy_level = 0
+        for i in range(3, 0, -1):
+            if len(p) > i:
+                if p[i] == "parent":
+                    hierarchy_level += 1
+
+        # sanity
+        if primary_resource_type == strings.USER_RESOURCE and hierarchy_level > 2:
+            raise ValueError("parent.parent.parent is invalid for a resource of type '{}'".format(str(strings.USER_RESOURCE)))
+        elif primary_resource_type == strings.RESOURCE_TYPE_WORKSPACE_SERVICE and hierarchy_level > 1:
+            raise ValueError("parent.parent is invalid for a resource of type '{}'".format(str(strings.RESOURCE_TYPE_WORKSPACE_SERVICE)))
+        elif primary_resource_type == strings.RESOURCE_TYPE_WORKSPACE and hierarchy_level > 0:
+            raise ValueError("parent is invalid for a resource of type '{}'".format(str(strings.RESOURCE_TYPE_WORKSPACE)))
+        elif primary_resource_type == strings.RESOURCE_TYPE_SHARED_SERVICE and hierarchy_level > 0:
+            raise ValueError("parent is invalid for a resource of type '{}'".format(str(strings.RESOURCE_TYPE_SHARED_SERVICE)))
+
+        if hierarchy_level == 2:
+            if primary_resource_type == strings.USER_RESOURCE:
+                dict_to_use = primary_parent_ws_dict
+                del p[2]
+                del p[1]
+        elif hierarchy_level == 1:
+            if primary_resource_type == strings.USER_RESOURCE:
+                dict_to_use = primary_parent_ws_svc_dict
+            elif primary_resource_type == strings.RESOURCE_TYPE_WORKSPACE_SERVICE:
+                dict_to_use = primary_parent_ws_dict
+            del p[1]
+        else:
+            dict_to_use = primary_resource_dict
+
+        prop_to_get = dict_to_use
+        for i in range(1, len(p)):
+            # instead of failing, if the value is not found, return empty string. Used for backward compatability
+            if p[i] not in prop_to_get:
```

Comment:
nit: I would suggest for readability put p[i] in a var since it is being used multiple time